File: pywcs/sip.py
# Copyright (C) 2008 Association of Universities for Research in Astronomy (AURA)

# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:

#     1. Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.

#     2. Redistributions in binary form must reproduce the above
#       copyright notice, this list of conditions and the following
#       disclaimer in the documentation and/or other materials provided
#       with the distribution.

#     3. The name of AURA and its representatives may not be used to
#       endorse or promote products derived from this software without
#       specific prior written permission.

# THIS SOFTWARE IS PROVIDED BY AURA ``AS IS'' AND ANY EXPRESS OR IMPLIED
# WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL AURA BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
# OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
# TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
# USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
# DAMAGE.

__docformat__ = "epytext"

import pyfits
import numpy
from pywcs import WCS

class SIP(WCS):
    """
    An extended WCS that supports Simple (or Spitzer) Imaging
    Polynomial (SIP) distortion coefficients.  In addition to all of
    the functionality of regular WCS objects, SIP objects also provide
    L{pix2foc} and L{foc2pix} methods to correct or apply the SIP
    distortion.

    @param header: A PyFITS header object or a string containing the raw
        FITS header data.
    @type header: PyFITS header object or string

    @param relax: Degree of permissiveness:
        - C{False}: Recognize only FITS keywords defined by the
          published WCS standard.
        - C{True}: Admit all recognized informal extensions of the
          WCS standard.
    @type relax: bool

    @raises MemoryError: Memory allocation failed.
    @raises ValueError: Invalid key.
    @raises ValueError: Header does not contain SIP information.
    @raises KeyError: Key not found in FITS header.
    """
    def __init__(self, header, relax=False):
        assert isinstance(header, pyfits.NP_pyfits.Header)

        if header.has_key("A_ORDER"):
            m = int(header["A_ORDER"])
            self._a = numpy.zeros((m+1, m+1))
            for i in range(m+1):
                for j in range(m-i+1):
                    self._a[i, j] = header.get("A_%d_%d" % (i, j), 0.0)

            if not header.has_key("B_ORDER"):
                raise ValueError(
                    "A_ORDER provided without corresponding B_ORDER "
                    "keyword for SIP distortion")

            m = int(header["B_ORDER"])
            self._b = numpy.zeros((m+1, m+1))
            for i in range(m+1):
                for j in range(m-i+1):
                    self._b[i, j] = header.get("B_%d_%d" % (i, j), 0.0)
        elif header.has_key("B_ORDER"):
            raise ValueError(
                "B_ORDER provided without corresponding A_ORDER "
                "keyword for SIP distortion")
        else:
            self._a = None
            self._b = None

        if header.has_key("AP_ORDER"):
            m = int(header["AP_ORDER"])
            self._ap = numpy.zeros((m+1, m+1))
            for i in range(m+1):
                for j in range(m-i+1):
                    self._ap[i, j] = header.get("AP_%d_%d" % (i, j), 0.0)

            if not header.has_key("BP_ORDER"):
                raise ValueError(
                    "AP_ORDER provided without corresponding BP_ORDER "
                    "keyword for SIP distortion")

            m = int(header["BP_ORDER"])
            self._bp = numpy.zeros((m+1, m+1))
            for i in range(m+1):
                for j in range(m-i+1):
                    self._bp[i, j] = header.get("BP_%d_%d" % (i, j), 0.0)
        elif header.has_key("BP_ORDER"):
            raise ValueError(
                "BP_ORDER provided without corresponding AP_ORDER "
                "keyword for SIP distortion")
        else:
            self._ap = None
            self._bp = None

        WCS.__init__(self, header, key=" ", relax=relax)

    def foc2pix(self, x, y=None):
        """
        Convert focal plane coordinates to pixel coordinates.

        B{Pixel coordinates are zero-based.}

        @param x: Array of I{y} focal plane coordinates, or if L{y} is
            not provided, a 2-dimensional array of both I{x}- and
            I{y}-coordinates.
        @type x: array of double

        @param y: Array of I{y} focal plane coordinates.  If provided,
            L{x} must be 1-dimensional and the same length as L{y}.
        @type y: array of double or None

        @return: If both L{x} and {y} were provided, a 2-tuple of
            arrays, where the first element is I{u} pixel coordinates
            and the second element is I{v} pixel coordinates.
            Otherwise, a single 2D array containing both I{u} and
            I{v}.

        @raises AssertionError: C{len(x) == len(y)}
        """
        if self._ap is None or self._bp is None:
            raise ValueError(
                "This object can not perform foc2pix conversion, since "
                "no AP or BP information was provided in the FITS header.")

        # If a single array was passed in, split it
        if y is None:
            y = x[:, 1]
            x = x[:, 0]
            numarrays = 1
        else:
            numarrays = 2
        assert len(x) == len(y)

        # Convert base-zero to base-one
        x += 1.0
        y += 1.0

        length = len(x)
        ap = self.ap
        bp = self.bp
        m = self.ap_order
        n = self.bp_order

        temp_x = x - self.crpix[0]
        temp_y = y - self.crpix[1]

        s = numpy.zeros((m + 1, length))
        for j in range(m+1):
            s[j] = ap[m-j, j]
            for k in range(j-1, -1, -1):
                s[j] = temp_y * s[j] + ap[m-j, k]

        sum = s[0]
        for i in range(m, 0, -1):
            sum = temp_x * sum + s[m-i+1]

        u = sum

        s = numpy.zeros((n + 1, length))
        for j in range(n+1):
            s[j] = bp[n-j, j]
            for k in range(j-1, -1, -1):
                s[j] = temp_y * s[j] + bp[n-j, k]

        sum = s[0]
        for i in range(n, 0, -1):
            sum = temp_x * sum + s[n-i+1]

        v = sum

        u += x
        v += y

        # Convert base-one back to base-zero
        x -= 1.0
        y -= 1.0

        if numarrays == 1:
            return numpy.hstack((u.reshape((length, 1)),
                                 v.reshape((length, 1))))
        return u, v

    def pix2foc(self, u, v=None):
        """
        Convert pixel coordinates to focal plane coordinates.

        B{Pixel coordinates are zero-based.}

        @param u: Array of I{u} pixel coordinates, or if L{v} is
            not provided, a 2-dimensional array of both I{u}- and
            I{v}-coordinates.
        @type u: array of double

        @param v: Array of I{v} pixel coordinates.  If provided,
            L{u} must be 1-dimensional and the same length as L{v}.
        @type v: array of double or None

        @return: If both L{u} and {v} were provided, a 2-tuple of
            arrays, where the first element is I{x} focal plane
            coordinates and the second element is I{y} focal plane
            coordinates.  Otherwise, a single 2D array containing both
            I{x} and I{y}.

        @raises AssertionError: C{len(u) == len(v)}
        """
        if self._a is None or self._b is None:
            raise ValueError(
                "This object can not perform foc2pix conversion, since "
                "no A or B information was provided in the FITS header.")

        # If a single array was passed in, split it
        if v is None:
            v = u[:, 1]
            u = u[:, 0]
            numarrays = 1
        else:
            numarrays = 2
        assert len(u) == len(v)

        length = len(u)
        a = self.a
        b = self.b
        m = self.a_order
        n = self.b_order
        temp_u = u - self.crpix[0]
        temp_v = v - self.crpix[1]

        s = numpy.zeros((m + 1, length))
        for j in range(m+1):
            s[j] = a[m-j, j]
            for k in range(j-1, -1, -1):
                s[j] = (temp_v * s[j]) + a[m-j, k]

        sum = s[0]
        for i in range(m, 0, -1):
            sum = temp_u * sum + s[m-i+1]

        x = sum

        s = numpy.zeros((m+1, length))
        for j in range(n+1):
            s[j] = b[n-j, j]
            for k in range(j-1, -1, -1):
                s[j] = (temp_v * s[j]) + b[n-j, k]

        sum = s[0]
        for i in range(n, 0, -1):
            sum = temp_u * sum + s[n-i+1]

        y = sum

        x += u
        y += v

        # Convert base-one to base-zero
        x -= 1.0
        y -= 1.0

        if numarrays == 1:
            return numpy.hstack((x.reshape((length, 1)),
                                 y.reshape((length, 1))))
        return x, y

    #@property
    def get_a_order(self):
        if self._a is None:
            raise ValueError("No A_ORDER provided in the FITS header.")
        return self._a.shape[0] - 1
    a_order = property(
        get_a_order,
        doc = """
              The A_ORDER value.
              @type: int
              """)

    #@property
    def get_b_order(self):
        if self._b is None:
            raise ValueError("No B_ORDER provided in the FITS header.")
        return self._b.shape[0] - 1
    b_order = property(
        get_b_order,
        doc = """
              The B_ORDER value.
              @type: int
              """)

    #@property
    def get_ap_order(self):
        if self._ap is None:
            raise ValueError("No AP_ORDER provided in the FITS header.")
        return self._ap.shape[0] - 1
    ap_order = property(
        get_ap_order,
        doc = """
              The AP_ORDER value.
              @type: int
              """)

    #@property
    def get_bp_order(self):
        if self._bp is None:
            raise ValueError("No BP_ORDER provided in the FITS header.")
        return self._bp.shape[0] - 1
    bp_order = property(
        get_bp_order,
        doc = """
              The BP_ORDER value.
              @type: int
              """)

    #@property
    def get_a(self):
        if self._a is None:
            raise ValueError("No A* provided in the FITS header.")
        return self._a
    a = property(
        get_a,
        doc = """
              The A coefficient array.
              @type: array[a_order+1][a_order+1] of double
              """)

    #@property
    def get_b(self):
        if self._b is None:
            raise ValueError("No B* provided in the FITS header.")
        return self._b
    b = property(
        get_b,
        doc = """
              The B coefficient array.
              @type: array[b_order+1][b_order+1] of double
              """)

    #@property
    def get_ap(self):
        if self._ap is None:
            raise ValueError("No AP* provided in the FITS header.")
        return self._ap
    ap = property(
        get_ap,
        doc = """
              The AP coefficient array.
              @type: array[ap_order+1][ap_order+1] of double
              """)

    #@property
    def get_bp(self):
        if self._bp is None:
            raise ValueError("No BP* provided in the FITS header.")
        return self._bp
    bp = property(
        get_bp,
        doc = """
              The BP coefficient array.
              @type: array[bp_order+1][bp_order+1] of double
              """)

File: src/sip.c
/*
Copyright (C) 2008 Association of Universities for Research in Astronomy (AURA)

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    1. Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    2. Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    3. The name of AURA and its representatives may not be used to
      endorse or promote products derived from this software without
      specific prior written permission.

THIS SOFTWARE IS PROVIDED BY AURA ``AS IS'' AND ANY EXPRESS OR IMPLIED
WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL AURA BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
DAMAGE.
*/

/*
 Author: Michael Droettboom
         mdroe@stsci.edu
*/

#include "sip.h"

#include <assert.h>
#include <stdlib.h>
#include <string.h>

void
sip_clear(sip_t* sip) {
  assert(sip);

  sip->a_order = 0;
  sip->a = NULL;
  sip->b_order = 0;
  sip->b = NULL;
  sip->ap_order = 0;
  sip->ap = NULL;
  sip->bp_order = 0;
  sip->bp = NULL;
  sip->crpix[0] = 0.0;
  sip->crpix[1] = 0.0;
  sip->scratch = NULL;
}

int
sip_init(
    sip_t* sip,
    const unsigned int a_order, const double* a,
    const unsigned int b_order, const double* b,
    const unsigned int ap_order, const double* ap,
    const unsigned int bp_order, const double* bp,
    const double crpix[2]) {
  unsigned int a_size       = 0;
  unsigned int b_size       = 0;
  unsigned int ap_size      = 0;
  unsigned int bp_size      = 0;
  unsigned int scratch_size = 0;
  int          status       = 0;

  assert(sip);
  sip_clear(sip);

  /* We we have one of A/B or AP/BP, we must have both. */
  if (((a == NULL) ^ (b == NULL)) ||
      ((ap == NULL) ^ (bp == NULL))) {
    return 6;
  }

  if (a != NULL) {
    sip->a_order = a_order;
    a_size = (a_order + 1) * (a_order + 1) * sizeof(double);
    sip->a = malloc(a_size);
    if (sip->a == NULL) {
      status = 2;
      goto exit;
    }
    memcpy(sip->a, a, a_size);
    if (a_order > scratch_size) {
      scratch_size = a_order;
    }

    sip->b_order = b_order;
    b_size = (b_order + 1) * (b_order + 1) * sizeof(double);
    sip->b = malloc(b_size);
    if (sip->b == NULL) {
      status = 2;
      goto exit;
    }
    memcpy(sip->b, b, b_size);
    if (b_order > scratch_size) {
      scratch_size = b_order;
    }
  }

  if (ap != NULL) {
    sip->ap_order = ap_order;
    ap_size = (ap_order + 1) * (ap_order + 1) * sizeof(double);
    sip->ap = malloc(ap_size);
    if (sip->ap == NULL) {
      status = 2;
      goto exit;
    }
    memcpy(sip->ap, ap, ap_size);
    if (ap_order > scratch_size) {
      scratch_size = ap_order;
    }

    sip->bp_order = bp_order;
    bp_size = (bp_order + 1) * (bp_order + 1) * sizeof(double);
    sip->bp = malloc(bp_size);
    if (sip->bp == NULL) {
      status = 2;
      goto exit;
    }
    memcpy(sip->bp, bp, bp_size);
    if (bp_order > scratch_size) {
      scratch_size = bp_order;
    }
  }

  if (scratch_size > 0) {
    scratch_size = (scratch_size + 1) * sizeof(double);
    sip->scratch = malloc(scratch_size);
    if (sip->scratch == NULL) {
      status = 2;
      goto exit;
    }
  }

  sip->crpix[0] = crpix[0];
  sip->crpix[1] = crpix[1];

 exit:
  if (status != 0) {
    sip_free(sip);
  }

  return status;
}

void
sip_free(sip_t* sip) {
  free(sip->a);
  sip->a = NULL;
  free(sip->b);
  sip->b = NULL;
  free(sip->ap);
  sip->ap = NULL;
  free(sip->bp);
  sip->bp = NULL;
  free(sip->scratch);
  sip->scratch = NULL;
}

static inline double
lu(
    const unsigned int order,
    const double* matrix,
    const int x,
    const int y) {
  int index;
  assert(x >= 0 && x <= order);
  assert(y >= 0 && y <= order);
  index = x * (order + 1) + y;
  assert(index >= 0 && index < (order + 1) * (order + 1));

  return matrix[index];
}

static int
sip_compute(
    const unsigned int naxes,
    const unsigned int nelem,
    const unsigned int m,
    const double* a,
    const unsigned int n,
    const double* b,
    const double crpix[2],
    double* tmp,
    const double* input /* [NAXES][nelem] */,
    double* output /* [NAXES][nelem] */) {
  int i, j, k;
  double x, y, tmp_x, tmp_y;
  double sum;

  assert(a);
  assert(b);
  assert(crpix);
  assert(tmp);
  assert(input);
  assert(output);

  /* Avoid segfaults */
  if (input == NULL || output == NULL || tmp == NULL || crpix == NULL) {
    return 1;
  }

  /* If we have one, we must have both... */
  if ((a == NULL) ^ (b == NULL)) {
    return 6;
  }

  /* If no distortion, just copy values */
  if (a == NULL /* && b == NULL ... implied */) {
    for (i = 0; i < nelem; ++i) {
      output[i << 1] = input[i << 1];
      output[(i << 1) + 1] = input[(i << 1) + 1];
    }
    return 0;
  }

  for (i = 0; i < nelem; ++i) {
    x = input[i << 1];
    y = input[(i << 1) + 1];

    tmp_x = x - crpix[0];
    tmp_y = y - crpix[1];

    for (j = 0; j <= m; ++j) {
      tmp[j] = lu(m, a, m-j, j);
      for (k = j-1; k >= 0; --k) {
        tmp[j] = (tmp_y * tmp[j]) + lu(m, a, m-j, k);
      }
    }

    /* Don't know why this loop is convoluted */
    sum = tmp[0];
    for (j = m; j > 0; --j) {
      sum = tmp_x * sum + tmp[m - j + 1];
    }
    output[i << 1] = sum + x;

    for (j = 0; j <= n; ++j) {
      tmp[j] = lu(n, b, n-j, j);
      for (k = j-1; k >= 0; --k) {
        tmp[j] = tmp_y * tmp[j] + lu(n, b, n-j, k);
      }
    }

    /* Don't know why this loop is convoluted */
    sum = tmp[0];
    for (j = n; j > 0; --j) {
      sum = tmp_x * sum + tmp[n - j + 1];
    }
    output[(i << 1) + 1] = sum + y;
  }

  return 0;
}

int
sip_pix2foc(
    const sip_t* sip,
    const unsigned int naxes,
    const unsigned int nelem,
    const double* pix /* [NAXES][nelem] */,
    double* foc /* [NAXES][nelem] */) {
  if (sip == NULL) {
    return 1;
  }

  return sip_compute(naxes, nelem,
                     sip->a_order, sip->a,
                     sip->b_order, sip->b,
                     sip->crpix,
                     (double *)sip->scratch,
                     pix, foc);
}

int
sip_foc2pix(
    const sip_t* sip,
    const unsigned int naxes,
    const unsigned int nelem,
    const double* foc /* [NAXES][nelem] */,
    double* pix /* [NAXES][nelem] */) {
  if (sip == NULL) {
    return 1;
  }

  return sip_compute(naxes, nelem,
                     sip->ap_order, sip->ap,
                     sip->bp_order, sip->bp,
                     sip->crpix,
                     (double *)sip->scratch,
                     foc, pix);
}

